packages/common/src/hardhat/tasks/artifacts.js

```diff
@@ -0,0 +1,206 @@
+const path = require("path");
+const fs = require("fs");
+const uniqBy = require("lodash.uniqby");
+
+const { task, types } = require("hardhat/config");
+
+function removeFileIfExists(filename) {
+  try {
+    fs.unlinkSync(filename);
+  } catch (e) {
+    // do nothing.
+  }
+}
+
+async function getArtifactPathList(hre, relativeTo) {
+  const artifactPaths = await hre.artifacts.getArtifactPaths();
+
+  // Generate a unique list of artifacts and paths to them. Unique is necessary because there are some redundantly
+  // named contracts.
+  return uniqBy(
+    artifactPaths.map((artifactPath) => ({
+      contractName: path.basename(artifactPath).split(".")[0],
+      relativePath: `./${path.relative(path.dirname(relativeTo), artifactPath)}`,
+    })),
+    "contractName"
+  );
+}
+
+function getCorePath(hre, relativeTo) {
+  const artifactPath = hre.config.paths.artifacts ? path.join(hre.config.paths.artifacts, "../") : "./";
+  return `./${path.relative(relativeTo, artifactPath)}`;
+}
+
+function getAddressesMap(hre) {
+  // Generate a map of name => chain id => address.
+  const networksPath = path.join(getCorePath(hre, "./"), "networks");
+  const dirs = fs.readdirSync(networksPath);
+  const addresses = {};
+  for (const dir of dirs) {
+    const chainId = parseInt(dir.split(".")[0]);
+    const deployments = JSON.parse(fs.readFileSync(path.join(networksPath, dir), "utf8"));
+
+    // Loop over the deployments in the file and save each one.
+    for (const { contractName, address, deploymentName } of deployments) {
+      // If deploymentName isn't specified, use contractName.
+      const name = deploymentName ? deploymentName : contractName;
+      if (!addresses[name]) {
+        addresses[name] = {};
+      }
+      addresses[name][chainId] = address;
+    }
+  }
+  return addresses;
+}
+
+task("generate-contracts-frontend", "Generate typescipt for the contracts-frontend package")
```

Comment:
in this format, if I understand correctly, the artifacts still contain the bytecode of the contract. is this something we'd like to consider removing? I guess however that if you use the `typechain` exports then we'll end up tree-shaking away all the artifacts that contain bytecode so it amounts to just having the interfaces anyway.

Reply:
Yeah, the intention is that we make bytecode accessible but tree-shake it away when it isn't used.

I spent some time testing this and it looks like the tree shaking works as expected. I did have to copy parts of the artifacts into the ts source to get that to work, though.